rotkehlchen/chain/ethereum/adex/adex.py

```diff
@@ -0,0 +1,352 @@
+from collections import defaultdict
+from functools import partial
+from typing import TYPE_CHECKING, Any, DefaultDict, Dict, List, Optional, Union, overload
+
+from eth_typing.evm import ChecksumAddress
+from eth_utils import to_checksum_address
+from eth_utils.typing import HexAddress, HexStr
+from gevent.lock import Semaphore
+from typing_extensions import Literal
+
+from rotkehlchen.accounting.structures import Balance
+from rotkehlchen.assets.asset import EthereumToken
+from rotkehlchen.chain.ethereum.graph import GRAPH_QUERY_LIMIT, Graph, format_query_indentation
+from rotkehlchen.chain.ethereum.utils import generate_address_via_create2
+from rotkehlchen.errors import RemoteError
+from rotkehlchen.fval import FVal
+from rotkehlchen.inquirer import Inquirer
+from rotkehlchen.premium.premium import Premium
+from rotkehlchen.typing import ChecksumEthAddress, Price, Timestamp
+from rotkehlchen.user_messages import MessagesAggregator
+from rotkehlchen.utils.interfaces import EthereumModule
+
+from .graph import BONDS_QUERY, UNBOND_REQUESTS_QUERY, UNBONDS_QUERY
+from .typing import Bond, PoolBalance, Unbond, UnbondRequest
+from .utils import (
+    ADX_AMOUNT_MANTISSA,
+    CREATE2_SALT,
+    IDENTITY_FACTORY_ADDR,
+    IDENTITY_PROXY_INIT_CODE,
+    POOL_ID_POOL_NAME,
+)
+
+if TYPE_CHECKING:
+    from rotkehlchen.chain.ethereum.manager import EthereumManager
+    from rotkehlchen.db.dbhandler import DBHandler
+
+
+class Adex(EthereumModule):
+    """AdEx integration module
+
+    AdEx subgraph:
+    https://github.com/samparsky/adex_subgraph
+    """
+    def __init__(
+            self,
+            ethereum_manager: 'EthereumManager',
+            database: 'DBHandler',
+            premium: Optional[Premium],
+            msg_aggregator: MessagesAggregator,
+    ) -> None:
+        self.ethereum = ethereum_manager
+        self.database = database
+        self.premium = premium
+        self.msg_aggregator = msg_aggregator
+        self.trades_lock = Semaphore()
+        try:
+            self.graph: Optional[Graph] = Graph(
+                'https://api.thegraph.com/subgraphs/name/adexnetwork/adex-protocol',
+            )
+        except RemoteError as e:
+            self.graph = None
+            self.msg_aggregator.add_error(
+                f'Could not initialize the AdEx subgraph due to {str(e)}. '
+                f'All AdEx balances and historical queries are not functioning until this is fixed. '  # noqa: E501
+                f'Probably will get fixed with time. If not report it to Rotkis support channel ',
+            )
+
+    @staticmethod
+    def _calculate_adex_balances(
```

Comment:
This is the function that would require a unit test imho.

The implementation kind of replicates [this logic](https://github.com/AdExNetwork/adex-staking/blob/af735d4e141cd160db91a97086582e34c8877e20/src/actions/actions.js#L353), but instead of updating the status of a bond/unbond/unbond_requested event, it just appends the bonds that have not been "unbonded" or "requested to be unbond".

I need to ask/check whether the "slashing" logic currently applies or not.

Reply:
I mentioned it already that we should not care about slashing events from a historical perspective at the moment, they can come in another iteration.

But if they need to be taken into account for the balances calculation we don't have much of a choice.